### ISSUE TITLE
Adding a new property to ConfigBuilder

### ISSUE THREAD

**Issue by theonepichael** [victools/jsonschema-generator#319]
I would like to add a "$schema" field to the "properties" object on the schema returned by the schema generator. I want to avoid adding this field to the model object classes themselves.

I want this:
```
{
  "$schema" : "https://json-schema.org/draft/2020-12/schema",
  "type" : "object",
  "properties" : {
    "uuid" : {
      "type" : "string",
      "description" : "Universally unique identifier",
      "format" : "uuid"
    }
  },
  "required" : [ "uuid" ],
  "additionalProperties" : false
}
```
To be this:
```
{
  "$schema" : "https://json-schema.org/draft/2020-12/schema",
  "type" : "object",
  "properties" : {
    "uuid" : {
      "type" : "string",
      "description" : "Universally unique identifier",
      "format" : "uuid"
    },
    "$schema": {
         type": "string"
     }
  },
  "required" : [ "uuid" ],
  "additionalProperties" : false
}
```

**Comment by CarstenWickner** [comment]
Hi @theonepichael,

The easiest way to do this would probably be to just manipulate the schema for each type (or just the ones you care about) via a "[Type Attribute Override](https://victools.github.io/jsonschema-generator/#type-attribute-overrides)", e.g. like the one below:
```java
config.forTypesInGeneral().withTypeAttributeOverride((schema, scope, context) -> schema
        .with(context.getKeyword(SchemaKeyword.TAG_PROPERTIES))
        .set(context.getKeyword(SchemaKeyword.TAG_SCHEMA),
                context.createStandardDefinitionReference(scope.getContext().resolve(String.class), null)));
```

**Comment by theonepichael** [comment]
When calling ```.with(context.getKeyword(SchemaKeyword.TAG_PROPERTIES))```, I get the following error: 
```
Invalid input: JSON Pointer expression must start with '/': "properties"
```

I appreciate the response.

**Comment by CarstenWickner** [comment]
Hi @theonepichael,

Which Jackson library version are you using?

* The `context.getKeyword(SchemaKeyword.TAG_PROPERTIES)` is just a way to avoid hard-coding `"properties"` in your code, which appears to be working correctly.
* The `ObjectNode.with("properties")` call should be returning a fresh jackson `ObjectNode`.
I don't see an obvious error.

Please provide a full stack trace, in order to get a better feeling where that error is being raised.

**Comment by theonepichael** [comment]
Below is the stack trace:
```
Exception in thread "main" java.lang.NullPointerException: Cannot invoke "io.swagger.v3.oas.annotations.media.Schema.minimum()" because the return value of "com.github.victools.jsonschema.generator.FieldScope.getAnnotation(java.lang.Class)" is null
        at com.trainer.jsonschemagenerator.JsonschemaGeneratorApplication.lambda$4(JsonschemaGeneratorApplication.java:85)
        at com.github.victools.jsonschema.generator.SchemaGeneratorTypeConfigPart.lambda$getFirstDefinedValue$0(SchemaGeneratorTypeConfigPart.java:50)
        at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
        at java.base/java.util.ArrayList$ArrayListSpliterator.tryAdvance(ArrayList.java:1602)
        at java.base/java.util.stream.ReferencePipeline.forEachWithCancel(ReferencePipeline.java:129)
        at java.base/java.util.stream.AbstractPipeline.copyIntoWithCancel(AbstractPipeline.java:527)
        at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:513)
        at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
        at java.base/java.util.stream.FindOps$FindOp.evaluateSequential(FindOps.java:150)
        at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
        at java.base/java.util.stream.ReferencePipeline.findFirst(ReferencePipeline.java:647)
        at com.github.victools.jsonschema.generator.SchemaGeneratorTypeConfigPart.getFirstDefinedValue(SchemaGeneratorTypeConfigPart.java:52)
        at com.github.victools.jsonschema.generator.SchemaGeneratorTypeConfigPart.resolveNumberExclusiveMinimum(SchemaGeneratorTypeConfigPart.java:401)
        at com.github.victools.jsonschema.generator.impl.SchemaGeneratorConfigImpl.resolveNumberExclusiveMinimum(SchemaGeneratorConfigImpl.java:529)
        at com.github.victools.jsonschema.generator.impl.AttributeCollector.collectFieldAttributes(AttributeCollector.java:84)
        at com.github.victools.jsonschema.generator.impl.SchemaGenerationContextImpl.createFieldSchema(SchemaGenerationContextImpl.java:632)
        at com.github.victools.jsonschema.generator.impl.SchemaGenerationContextImpl.populateFieldSchema(SchemaGenerationContextImpl.java:608)
        at com.github.victools.jsonschema.generator.impl.SchemaGenerationContextImpl.generateObjectDefinition(SchemaGenerationContextImpl.java:459)
        at com.github.victools.jsonschema.generator.impl.SchemaGenerationContextImpl.addSubtypeReferencesInDefinition(SchemaGenerationContextImpl.java:366)
        at com.github.victools.jsonschema.generator.impl.SchemaGenerationContextImpl.traverseGenericType(SchemaGenerationContextImpl.java:339)
        at com.github.victools.jsonschema.generator.impl.SchemaGenerationContextImpl.traverseGenericType(SchemaGenerationContextImpl.java:276)
        at com.github.victools.jsonschema.generator.impl.SchemaGenerationContextImpl.parseType(SchemaGenerationContextImpl.java:97)
        at com.github.victools.jsonschema.generator.SchemaBuilder.createSchemaForSingleType(SchemaBuilder.java:111)
        at com.github.victools.jsonschema.generator.SchemaBuilder.createSingleTypeSchema(SchemaBuilder.java:56)
        at com.github.victools.jsonschema.generator.SchemaGenerator.generateSchema(SchemaGenerator.java:60)
        at com.trainer.jsonschemagenerator.JsonschemaGeneratorApplication.toJsonSchema(JsonschemaGeneratorApplication.java:96)
        at com.trainer.jsonschemagenerator.JsonschemaGeneratorApplication.loadClass(JsonschemaGeneratorApplication.java:62)
        at com.trainer.jsonschemagenerator.JsonschemaGeneratorApplication.main(JsonschemaGeneratorApplication.java:49
```

**Comment by theonepichael** [comment]
A follow up:
I want to generate a schema that looks like the following:

{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "title": "Example",
  "anyOf": [
    {
      "required": [
        "foo"
      ]
    },
    {
      "required": [
        "bar"
      ]
    },
    {
      "required": [
        "baz"
      ]
    },
    {
      "required": [
        "fizz"
      ]
    },
    {
      "required": [
        "buzz"
      ]
    }
  ],
  "properties": {
    "$schema": {
      "type": "string"
    },
    "foo": {
      "description": "foo",
      "type": "number",
      "minimum": 0.0,
      "maximum": 360.0
    },
    "bar": {
      "description": "bar",
      "type": "number",
      "minimum": 0.0
    },
    "baz": {
      "description": "baz",
      "type": "number"
    },
    "fizz": {
      "description": "fizz",
      "type": "number",
      "minimum": 0.0
    },
    "buzz": {
      "description": "buzz,
      "type": "number",
      "minimum": 0.0
    }
  },
  "additionalProperties": false
}
One line explanation of the schema:
"The json can have any of the required properties, but if any of them are included then they must conform to 
their corresponding entry in 'properties'"

**Comment by CarstenWickner** [comment]
Hi @theonepichael,

I can't reproduce your exact error, but to make it work I had to add an extra filter in the `TypeAttributeOverride` configuration to avoid an endless loop.
My minimal working example is this:
```java
public class TypeAttributeOverride319 {

    public static void main(String[] args) {
        SchemaGeneratorConfigBuilder configBuilder = new SchemaGeneratorConfigBuilder(SchemaVersion.DRAFT_2020_12, OptionPreset.PLAIN_JSON);
        configBuilder.forTypesInGeneral().withTypeAttributeOverride((schema, scope, context) -> {
            if (scope.getType().getErasedType() == TestType.class) {
                schema.with(context.getKeyword(SchemaKeyword.TAG_PROPERTIES))
                    .set(context.getKeyword(SchemaKeyword.TAG_SCHEMA),
                            context.createStandardDefinition(context.getTypeContext().resolve(String.class), null));
            }
        });
        SchemaGeneratorConfig config = configBuilder.build();
        SchemaGenerator generator = new SchemaGenerator(config);
        JsonNode jsonSchema = generator.generateSchema(TestType.class);

        System.out.println(jsonSchema.toPrettyString());
    }

    public static class TestType {
        public String otherProperty;
    }
}
```
This produces:
```json
{
  "$schema" : "https://json-schema.org/draft/2020-12/schema",
  "type" : "object",
  "properties" : {
    "otherProperty" : {
      "type" : "string"
    },
    "$schema" : {
      "type" : "string"
    }
  }
}
```

----

Regarding your follow-up question:
* If the `$schema` property was always required, you could just set the `minProperties` to `2` in order to ensure that at least one of the "normal" properties is present.
* Alternatively, you can do something as in the following example:

```java
public class AnyOfRequired319 {

    public static void main(String[] args) {
        SchemaGeneratorConfigBuilder configBuilder = new SchemaGeneratorConfigBuilder(SchemaVersion.DRAFT_7, OptionPreset.PLAIN_JSON);
        configBuilder.forTypesInGeneral()
                // use CustomDefinitionProvider to add anyOf that marks each individual property as required
                .withCustomDefinitionProvider(new AnyPropertyRequiredCustomDefinitionProvider(TestType.class));
        configBuilder
                // include JakartaValidationModule in order to pick up minimum/maximum constraints
                .with(new JakartaValidationModule())
                // include JacksonModule in order to pick up annotated descriptions
                .with(new JacksonModule());
        SchemaGeneratorConfig config = configBuilder.build();
        SchemaGenerator generator = new SchemaGenerator(config);
        JsonNode jsonSchema = generator.generateSchema(TestType.class);

        System.out.println(jsonSchema.toPrettyString());
    }

    private static class AnyPropertyRequiredCustomDefinitionProvider implements CustomDefinitionProviderV2 {

        private final Class<?> erasedTargetType;

        public AnyPropertyRequiredCustomDefinitionProvider(Class<?> erasedTargetType) {
            this.erasedTargetType = erasedTargetType;
        }

        @Override
        public CustomDefinition provideCustomSchemaDefinition(ResolvedType javaType, SchemaGenerationContext context) {
            if (javaType.getErasedType() != this.erasedTargetType) {
                return null;
            }
            ObjectNode schema = context.createStandardDefinition(javaType, this);
            ObjectNode propertiesNode = schema.with(context.getKeyword(SchemaKeyword.TAG_PROPERTIES));
            if (!propertiesNode.isEmpty()) {
                // wrap additional anyOf into allOf in order to avoid conflicting with potentially existing anyOf
                // if the allOf is unnecessary, it will be automatically removed in the end
                ArrayNode anyOfNode = schema.withArray(context.getKeyword(SchemaKeyword.TAG_ALLOF))
                            .addObject().withArray(context.getKeyword(SchemaKeyword.TAG_ANYOF));
                propertiesNode.fieldNames()
                        .forEachRemaining(fieldName -> anyOfNode.addObject()
                                .withArray(context.getKeyword(SchemaKeyword.TAG_REQUIRED))
                                .add(fieldName));
            }
            propertiesNode.set(context.getKeyword(SchemaKeyword.TAG_SCHEMA),
                        context.createStandardDefinition(context.getTypeContext().resolve(String.class), null));
            return new CustomDefinition(schema, CustomDefinition.DefinitionType.STANDARD, CustomDefinition.AttributeInclusion.YES);
        }
    }

    public static class TestType {
        @DecimalMin("0.0")
        @DecimalMax("360.0")
        @JsonPropertyDescription("foo")
        public double foo;
        @DecimalMin("0.0")
        @JsonPropertyDescription("bar")
        public double bar;
        @JsonPropertyDescription("baz")
        public double baz;
        @DecimalMin("0.0")
        @JsonPropertyDescription("fizz")
        public double fizz;
        @DecimalMin("0.0")
        @JsonPropertyDescription("buzz")
        public double buzz;
    }
}
```
This produces:
```json
{
  "$schema" : "http://json-schema.org/draft-07/schema#",
  "type" : "object",
  "properties" : {
    "bar" : {
      "type" : "number",
      "description" : "bar",
      "minimum" : 0.0
    },
    "baz" : {
      "type" : "number",
      "description" : "baz"
    },
    "buzz" : {
      "type" : "number",
      "description" : "buzz",
      "minimum" : 0.0
    },
    "fizz" : {
      "type" : "number",
      "description" : "fizz",
      "minimum" : 0.0
    },
    "foo" : {
      "type" : "number",
      "description" : "foo",
      "minimum" : 0.0,
      "maximum" : 360.0
    },
    "$schema" : {
      "type" : "string"
    }
  },
  "anyOf" : [ {
    "required" : [ "bar" ]
  }, {
    "required" : [ "baz" ]
  }, {
    "required" : [ "buzz" ]
  }, {
    "required" : [ "fizz" ]
  }, {
    "required" : [ "foo" ]
  } ]
}
```

I hope that helps. 😃

**Comment by theonepichael** [comment]
> 
Thanks for the response. Looks like the ```.with``` method has been deprecated:
```java
schema.with(context.getKeyword(SchemaKeyword.TAG_PROPERTIES))
```

The suggested alternative of the ```.withObject``` does not work either and I get a stack trace that is extremely similar to the one from my earlier post. Your advise is much appreciated.

Including pom.xml:

```xml
	<parent>
		<groupId>org.springframework.boot</groupId>
		<artifactId>spring-boot-starter-parent</artifactId>
		<version>3.0.4</version>
		<relativePath /> <!-- lookup parent from repository -->
	</parent>
	<properties>
		<java.version>17</java.version>
	</properties>
	<dependencies>
		<dependency>
			<groupId>org.springframework.boot</groupId>
			<artifactId>spring-boot-starter</artifactId>
		</dependency>

		<dependency>
			<groupId>org.springframework.boot</groupId>
			<artifactId>spring-boot-starter-test</artifactId>
			<scope>test</scope>
		</dependency>

		<dependency>
			<groupId>org.projectlombok</groupId>
			<artifactId>lombok</artifactId>
			<version>1.18.24</version>
		</dependency>

		<dependency>
			<groupId>com.github.victools</groupId>
			<artifactId>jsonschema-generator</artifactId>
			<version>4.28.0</version>
		</dependency>

		<dependency>
			<groupId>com.github.victools</groupId>
			<artifactId>jsonschema-module-jackson</artifactId>
			<version>4.28.0</version>
		</dependency>

		<dependency>
			<groupId>com.github.victools</groupId>
			<artifactId>jsonschema-module-swagger-2</artifactId>
			<version>4.21.0</version>
		</dependency>

		<dependency>
			<groupId>org.springdoc</groupId>
			<artifactId>springdoc-openapi-starter-webmvc-ui</artifactId>
			<version>2.0.2</version>
		</dependency>

	</dependencies>
```

**Comment by CarstenWickner** [comment]
Hi @theonepichael,

I've edited my previous answer to replace the usage of the deprecated line with this.
```java
ObjectNode propertiesNode = (ObjectNode) schema.putIfAbsent(context.getKeyword(SchemaKeyword.TAG_PROPERTIES), schema.objectNode());
```

If one assumes, that a `"properties"` element is always present, one could also just `get()` it:
```java
(ObjectNode) schema.get(context.getKeyword(SchemaKeyword.TAG_PROPERTIES))
```

**Comment by theonepichael** [comment]
Thanks for clearing that up. I have an additional follow up.

I want to produce a schema that looks like the following.
(WHERE I WANT TO GET TO):
```json
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "title": "Example",
  "required": [
    "alpha",
    "beta",
    "sigma"
  ],
  "properties": {
    "$schema": {
      "type": "string"
    },
    "alpha": {
      "description": "alpha",
      "type": "string"
    },
    "beta": {
      "$ref": "./BetaSchema.json"
    },
    "sigma": {
      "description": "sigma",
      "type": "number"
    },
    "omega": {
      "type": "array",
      "items": {
        "oneOf": [
          {
            "$ref": "./ThetaSchema.json"
          },
          {
            "$ref": "./TauSchema.json"
          }
        ]
      }
    }
  },
  "additionalProperties": false
}
```
One line explanation of the schema:
"The json must have all of the required properties, and if 'omega' is included, the array can only contain one of the $ref objects"


(WHERE I'M CURRENTLY AT) The current output is:
```json
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "title": "Example",
  "required": [
    "alpha",
    "beta",
    "sigma"
  ],
  "properties": {
    "$schema": {
      "type": "string"
    },
    "alpha": {
      "description": "alpha",
      "type": "string"
    },
    "beta": {
      "$ref": "./BetaSchema.json"
    },
    "sigma": {
      "description": "sigma",
      "type": "number"
    },
    "omega": {
      "type": "array",
      "items": {
        "anyOf": [ // should be "oneOf"
          {
            //theta schema is inlined -> should be "$ref" : "..."
          },
          {
            //tau schema is inlined -> should be "$ref" : "..."
          }
        ]
      }
    }
  },
  "additionalProperties": false
}
```

Note: The "Omega" class is the supertype of Theta and Tau. So only one of those subclass types is allowed in the array

```java
class Example {
    /// required properties omitted

  @Schema(name = "omega", type = "array", oneOf = {Theta.class, Tau.class})
  private List<Omega> omega;
} 

@Schema(
    title = "Theta",
    description = "Theta",
    additionalProperties = Schema.AdditionalPropertiesValue.FALSE,
    ref = "./ThetaSchema.json")
class Theta extends Omega {}

// Implementation omitted for brevity
@Schema(
    title = "Tau",
    description = "Tau",
    additionalProperties = Schema.AdditionalPropertiesValue.FALSE,
    ref = "./TauSchema.json")
class Tau extends Omega {}
```
What changes do I need to make to get to my desired output?

**Comment by CarstenWickner** [comment]
Hi @theonepichael,

It's okay to create a new question btw., no need to keep adding more things to the same issue. 😜

That being said, your annotation of the `List<Omega>` seems incorrect for what you want to achieve.
It should be this instead:
```java
    @ArraySchema(schema = @Schema(oneOf = {Theta.class, Tau.class}))
    private List<Omega> omega;
```

However, I can confirm that the `@Schema(ref = "...")` attribute is not being considered, when it is on the type.

I guess it should be added as a `CustomDefinitionProvider` inside the `Swagger2Module`, which you appear to be using. Something along the lines of what I did in another question related to replacing schemas with external references:
https://github.com/victools/jsonschema-generator/issues/246#issuecomment-1099666190

Until that is released, you can configure such a `CustomDefinitionProvider` yourself.

A complete example would be:
```java
public class ExternalRef319 {

    public static void main(String[] args) {
        SchemaGeneratorConfigBuilder configBuilder = new SchemaGeneratorConfigBuilder(SchemaVersion.DRAFT_7, OptionPreset.PLAIN_JSON);
        configBuilder.with(new Swagger2Module());
        configBuilder.forTypesInGeneral().withTypeAttributeOverride((schema, scope, context) -> {
            if (scope.getType().getErasedType() == Example.class) {
                ((ObjectNode) schema.putIfAbsent(context.getKeyword(SchemaKeyword.TAG_PROPERTIES), schema.objectNode()))
                    .set(context.getKeyword(SchemaKeyword.TAG_SCHEMA),
                            context.createStandardDefinition(context.getTypeContext().resolve(String.class), null));
            }
        });
        configBuilder.forTypesInGeneral().withCustomDefinitionProvider(new SchemaRefDefinitionProvider());
        SchemaGeneratorConfig config = configBuilder.build();
        SchemaGenerator generator = new SchemaGenerator(config);
        JsonNode jsonSchema = generator.generateSchema(Example.class);

        System.out.println(jsonSchema.toPrettyString());
    }

    private static class SchemaRefDefinitionProvider implements CustomDefinitionProviderV2 {

        private Class<?> mainType;

        @Override
        public CustomDefinition provideCustomSchemaDefinition(ResolvedType javaType, SchemaGenerationContext context) {
            Class<?> erasedType = javaType.getErasedType();
            if (this.mainType == null) {
                this.mainType = erasedType;
            }
            if (this.mainType == erasedType) {
                return null;
            }
            return Optional.ofNullable(erasedType.getAnnotation(Schema.class))
                    .map(Schema::ref)
                    .filter(ref -> !ref.isEmpty())
                    .map(ref -> context.getGeneratorConfig().createObjectNode()
                            .put(context.getKeyword(SchemaKeyword.TAG_REF), ref))
                    .map(schema -> new CustomDefinition(schema,
                            CustomDefinition.INLINE_DEFINITION,
                            CustomDefinition.INCLUDING_ATTRIBUTES))
                    .orElse(null);
        }

    }

    class Example {
        @Schema(description = "alpha")
        private String alpha;
        @Schema(ref = "./BetaSchema.json")
        private Object beta;
        @Schema(description = "sigma")
        private Double sigma;
        @ArraySchema(schema = @Schema(oneOf = {Theta.class, Tau.class}))
        private List<Omega> omega;
    }

    class Omega {
    }

    @Schema(
            title = "Theta",
            description = "Theta",
            additionalProperties = Schema.AdditionalPropertiesValue.FALSE,
            ref = "./ThetaSchema.json")
    class Theta extends Omega {
    }

    // Implementation omitted for brevity
    @Schema(
            title = "Tau",
            description = "Tau",
            additionalProperties = Schema.AdditionalPropertiesValue.FALSE,
            ref = "./TauSchema.json")
    class Tau extends Omega {
    }
}
```
The output of the above is:
```java
{
  "$schema" : "http://json-schema.org/draft-07/schema#",
  "type" : "object",
  "properties" : {
    "alpha" : {
      "type" : "string",
      "description" : "alpha"
    },
    "beta" : {
      "$ref" : "./BetaSchema.json"
    },
    "omega" : {
      "type" : "array",
      "items" : {
        "type" : "object",
        "oneOf" : [ {
          "$ref" : "./ThetaSchema.json",
          "title" : "Theta",
          "description" : "Theta"
        }, {
          "$ref" : "./TauSchema.json",
          "title" : "Tau",
          "description" : "Tau"
        } ]
      }
    },
    "sigma" : {
      "type" : "number",
      "description" : "sigma"
    },
    "$schema" : {
      "type" : "string"
    }
  }
}
```

**Comment by CarstenWickner** [comment]
HI @theonepichael,

I've just released [v4.29.0](https://github.com/victools/jsonschema-generator/releases/tag/v4.29.0), i.e., this is now part of the default `Swagger2Module` behavior.

Now, you just need to change the aforementioned annotation to the following (and update your `jsonschema-generator` dependency versions):
```java
    @ArraySchema(schema = @Schema(oneOf = {Theta.class, Tau.class}))
    private List<Omega> omega;
```
